### PR TITLE
implement fix pg size

### PR DESCRIPTION
**Pull request by Hooper9973**
## Description:

This pull request introduces the initial draft for fixing the PG size. The main work includes:

- Creating PGs based on a specified size.
- Selecting chunks from the PG's chunk heap when creating a shard.
- Related to recovery: Adapting `pg_info_superblk` to recover the state of chunkselector and the state of pg.

### CODE REVIEW

File: src/lib/homestore_backend/hs_pg_manager.cpp [hunk 1]
```diff
@@ -127,12 +136,19 @@ void HSHomeObject::on_create_pg_message_commit(int64_t lsn, sisl::blob const& he
         return;
     }
 
+    // select chunks for pg
+    const auto chunk_size = chunk_selector()->get_chunk_size();
+    const chunk_num_t num_chunk = sisl::round_up(pg_info.size, chunk_size) / chunk_size;
+    LOGINFOMOD(homeobject, "[Hooper] on create commit create pg {}, num_chunk {}, pg_info_size {}", pg_id, num_chunk, pg_info.size);
+
+    chunk_selector()->select_chunks_for_pg(pg_id, num_chunk);
```
Comment (xiaoxichen) [inline src/lib/homestore_backend/hs_pg_manager.cpp:144]: suggesting failing here. I think before L160 we can  *fail* a pg creation by not adding to map, following data write will fail as chunk_selector cannot give back hint,  append_log will fail

c.f https://github.com/eBay/HomeStore/blob/master/src/lib/replication/repl_dev/common.cpp#L91-L92

File: src/lib/homestore_backend/heap_chunk_selector.cpp [hunk 2]
```diff
@@ -148,6 +136,48 @@ csharedChunk HeapChunkSelector::select_specific_chunk(const chunk_num_t chunkID)
     return vchunk.get_internal_chunk();
 }
 
+csharedChunk HeapChunkSelector::select_specific_chunk(const pg_id_t pg_id, const chunk_num_t chunkID) {
```
[the remark this answers is not in the document]
Comment (xiaoxichen) [reply src/lib/homestore_backend/heap_chunk_selector.cpp:139]: also please add comment to clarify the original one is to-be-removed

File: src/lib/homestore_backend/hs_pg_manager.cpp (hunk 1 above)
Comment (Hooper9973) [reply src/lib/homestore_backend/hs_pg_manager.cpp]: Done

File: src/lib/homestore_backend/hs_pg_manager.cpp [hunk 3]
```diff
@@ -127,12 +136,19 @@ void HSHomeObject::on_create_pg_message_commit(int64_t lsn, sisl::blob const& he
         return;
     }
 
+    // select chunks for pg
+    const auto chunk_size = chunk_selector()->get_chunk_size();
+    const chunk_num_t num_chunk = sisl::round_up(pg_info.size, chunk_size) / chunk_size;
```
[the remark this answers is not in the document]
Comment (Hooper9973) [reply src/lib/homestore_backend/hs_pg_manager.cpp]: The same chunk size is already ensured by homogeneous disks, the same number of chunks is ensured by fixed pg size, and an identical layout will be guaranteed during the create shard process by writing `v_chunk_id` into `shard_info_superblk` to ensure followers select the same v_chunk.

File: src/lib/homestore_backend/heap_chunk_selector.cpp [hunk 4]
```diff
@@ -204,6 +258,145 @@ void HeapChunkSelector::build_per_dev_chunk_heap(const std::unordered_set< chunk
...
+    auto vchunk = VChunk(nullptr);
+    auto it = m_per_pg_heap.emplace(pg_id, std::make_shared< PerDevHeap >()).first;
+    auto v2r_vector = m_v2r_chunk_map.emplace(pg_id, std::make_shared< std::vector < chunk_num_t > >()).first->second;
+    auto r2v_map = m_r2v_chunk_map.emplace(pg_id, std::make_shared< ChunkIdMap >()).first->second;
+
+    auto& pg_heap = it->second;
+    std::scoped_lock lock(pdev_heap->mtx, pg_heap->mtx);
+    v2r_vector->reserve(num_chunk);
+    for (chunk_num_t i = 0; i < num_chunk; ++i) {
+        vchunk = pdev_heap->m_heap.top();
```
[the remark this answers is not in the document]
Comment (Hooper9973) [reply src/lib/homestore_backend/heap_chunk_selector.cpp:237]: Add RELEASE_ASSERT

File: src/lib/homestore_backend/hs_pg_manager.cpp [hunk 5]
```diff
@@ -127,12 +136,19 @@ void HSHomeObject::on_create_pg_message_commit(int64_t lsn, sisl::blob const& he
         return;
     }
 
+    // select chunks for pg
+    const auto chunk_size = chunk_selector()->get_chunk_size();
```
[the remark this answers is not in the document]
Comment (Hooper9973) [reply src/lib/homestore_backend/hs_pg_manager.cpp]: Done

File: src/lib/homestore_backend/hs_homeobject.hpp [hunk 6]
```diff
@@ -95,14 +97,24 @@ class HSHomeObject : public HomeObjectImpl {
...
-            memcpy(members, rhs.members, sizeof(pg_members) * num_members);
+
+            memcpy(get_pg_members(), rhs.get_pg_members(), sizeof(pg_members) * num_members);
+            memcpy(get_chunk_ids(), rhs.get_chunk_ids(), sizeof(homestore::chunk_num_t) * num_chunks);
             return *this;
         }
 
         void copy(pg_info_superblk const& rhs) { *this = rhs; }
+
+        pg_members* get_pg_members() { return reinterpret_cast<pg_members*>(data); }
```
[the remark this answers is not in the document]
Comment (Hooper9973) [reply src/lib/homestore_backend/hs_homeobject.hpp]: Done

File: src/lib/homestore_backend/heap_chunk_selector.cpp [hunk 7]
```diff
@@ -204,6 +258,145 @@ void HeapChunkSelector::build_per_dev_chunk_heap(const std::unordered_set< chunk
...
+        chunk_num_t r_chunk_id = vchunk.get_chunk_id();
+        v2r_vector->emplace_back(r_chunk_id);
+        r2v_map->emplace(r_chunk_id, v_chunk_id);
+    }
+    LOGINFOMOD(homeobject, "[Hooper] Successfully created pg {}, num_chunk {} ", pg_id, num_chunk);
+
+    return;
+}
+
+void HeapChunkSelector::persist_chunk_ids_into_pg_sb(pg_id_t pg_id, chunk_num_t num_chunk, homestore::chunk_num_t* pb_sb_chunk_ids) const {
```
[the remark this answers is not in the document]
Comment (Hooper9973) [reply src/lib/homestore_backend/heap_chunk_selector.cpp]: Indeed, that makes more sense. Done.

File: src/lib/homestore_backend/heap_chunk_selector.cpp [hunk 8]
```diff
@@ -204,6 +258,145 @@ void HeapChunkSelector::build_per_dev_chunk_heap(const std::unordered_set< chunk
...
+    std::shared_lock lock_guard(m_chunk_selector_mtx);
+    auto it = m_v2r_chunk_map.find(pg_id);
+    if (it == m_v2r_chunk_map.end()) {
+        LOGWARNMOD(homeobject, "Pg {} had never been created", pg_id);
+        return;
+    }
+    auto chunk_ids = it->second;
+    for (chunk_num_t i = 0; i < num_chunk; ++i) {
+        pb_sb_chunk_ids[i] = chunk_ids->at(i);
```
[the remark this answers is not in the document]
Comment (Hooper9973) [reply src/lib/homestore_backend/heap_chunk_selector.cpp]: The commented interface has been removed.

File: src/lib/homestore_backend/heap_chunk_selector.cpp [hunk 9]
```diff
@@ -204,6 +258,145 @@ void HeapChunkSelector::build_per_dev_chunk_heap(const std::unordered_set< chunk
...
+                            const std::pair< const uint32_t, std::shared_ptr< PerDevHeap > >& rhs) {
+                                return lhs.second->available_blk_count.load() < rhs.second->available_blk_count.load();
+                            });
+    auto& pdev_heap = most_avail_dev_it->second;
+    if (num_chunk > pdev_heap->size()) {
+        LOGWARNMOD(homeobject, "Pdev has no enough space to create pg {} with num_chunk {}", pg_id, num_chunk);
+        return;
+    }
+    auto vchunk = VChunk(nullptr);
+    auto it = m_per_pg_heap.emplace(pg_id, std::make_shared< PerDevHeap >()).first;
```
[the remark this answers is not in the document]
Comment (Hooper9973) [reply src/lib/homestore_backend/heap_chunk_selector.cpp]: Done

File: src/lib/homestore_backend/heap_chunk_selector.cpp [hunk 10]
```diff
@@ -204,6 +258,145 @@ void HeapChunkSelector::build_per_dev_chunk_heap(const std::unordered_set< chunk
     };
 }
 
+uint32_t HeapChunkSelector::get_chunk_size() const {
+    const auto& chunk = m_chunks.begin()->second;
+    auto vchunk = VChunk(chunk);
+    return vchunk.size();
+}
+
+void HeapChunkSelector::select_chunks_for_pg(pg_id_t pg_id, chunk_num_t num_chunk) {
```
[the remark this answers is not in the document]
Comment (Hooper9973) [reply src/lib/homestore_backend/heap_chunk_selector.cpp]: Done

File: src/lib/homestore_backend/heap_chunk_selector.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Hooper9973) [reply src/lib/homestore_backend/heap_chunk_selector.cpp:96]: The original one had been removed, the original one can only be called by test_heap_chunk_selector which will be refine in next commit.

File: src/lib/homestore_backend/heap_chunk_selector.cpp [hunk 11]
```diff
@@ -99,7 +87,7 @@ csharedChunk HeapChunkSelector::select_chunk(homestore::blk_count_t count, const
         avalableBlkCounter.fetch_sub(vchunk.available_blks());
         remove_chunk_from_defrag_heap(vchunk.get_chunk_id());
     } else {
-        LOGWARNMOD(homeobject, "No pdev found for pdev {}", pdevID);
+        LOGWARNMOD(homeobject, "No pg found for pg_id {}", pg_id);
```
[the remark this answers is not in the document]
Comment (Hooper9973) [reply src/lib/homestore_backend/heap_chunk_selector.cpp:90]: I think all vchunks in the heap are taken from `m_chunks`, and `m_chunks` is guaranteed by homestore to be not null, so we needn't assert.

File: src/lib/homestore_backend/hs_pg_manager.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (xiaoxichen) [reply src/lib/homestore_backend/hs_pg_manager.cpp]: yeah I am concerning about the chunk size ---- if mistaken was made, surface it up earlier.

File: src/lib/homestore_backend/heap_chunk_selector.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (xiaoxichen) [reply src/lib/homestore_backend/heap_chunk_selector.cpp:90]: then in which case we will go into this `else` path?

**Review by xiaoxichen — COMMENTED** [review]
in general LGTM

File: src/lib/homestore_backend/heap_chunk_selector.cpp [hunk 12]
```diff
@@ -204,6 +258,145 @@ void HeapChunkSelector::build_per_dev_chunk_heap(const std::unordered_set< chunk
...
 }
 
+uint32_t HeapChunkSelector::get_chunk_size() const {
+    const auto& chunk = m_chunks.begin()->second;
+    auto vchunk = VChunk(chunk);
+    return vchunk.size();
+}
+
+void HeapChunkSelector::select_chunks_for_pg(pg_id_t pg_id, chunk_num_t num_chunk) {
+    LOGINFOMOD(homeobject, "[Hooper] start created pg {}, num_chunk {} ", pg_id, num_chunk);
```
[the remark this answers is not in the document]
Comment (Hooper9973) [reply src/lib/homestore_backend/heap_chunk_selector.cpp]: done

File: src/lib/homestore_backend/heap_chunk_selector.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Hooper9973) [reply src/lib/homestore_backend/heap_chunk_selector.cpp:96]: test_heap_chunk_selector had been adapted

File: src/lib/homestore_backend/heap_chunk_selector.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (Hooper9973) [reply src/lib/homestore_backend/heap_chunk_selector.cpp]: no available chunks left for the specified PG

File: src/lib/homestore_backend/hs_pg_manager.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Hooper9973) [reply src/lib/homestore_backend/hs_pg_manager.cpp]: More comprehensive error handling and security defense code will be proposed in the Identical layout pr.

File: src/lib/homestore_backend/hs_pg_manager.cpp [hunk 13]
```diff
@@ -60,6 +60,15 @@ PGManager::NullAsyncResult HSHomeObject::_create_pg(PGInfo&& pg_info, std::set<
     auto pg_id = pg_info.id;
     if (auto lg = std::shared_lock(_pg_lock); _pg_map.end() != _pg_map.find(pg_id)) return folly::Unit();
 
+    const auto most_avail_num_chunks = chunk_selector()->most_avail_num_chunks();
+    const auto chunk_size = chunk_selector()->get_chunk_size();
+    const auto needed_num_chunks = sisl::round_up(pg_info.size, chunk_size) / chunk_size;
```
Comment (JacksonYao287) [inline src/lib/homestore_backend/hs_pg_manager.cpp:65]: I think it`s better to use round_down here

File: src/lib/homestore_backend/heap_chunk_selector.cpp [hunk 14]
```diff
@@ -59,31 +59,19 @@ csharedChunk HeapChunkSelector::select_chunk(homestore::blk_count_t count, const
...
-                             [](const std::pair< const uint32_t, std::shared_ptr< PerDevHeap > >& lhs,
-                                const std::pair< const uint32_t, std::shared_ptr< PerDevHeap > >& rhs) {
-                                 return lhs.second->available_blk_count.load() < rhs.second->available_blk_count.load();
-                             });
-        if (it == m_per_dev_heap.end()) {
-            LOGWARNMOD(homeobject, "No pdev found for new pg");
-            return nullptr;
-        }
-        pdevID = it->first;
+    // Temporary bypass using pdev_id_hint to represent pg_id_hint
```
Comment (JacksonYao287) [inline src/lib/homestore_backend/heap_chunk_selector.cpp:62]: if two pg reside in the same pdev, how to handle this case?

File: src/lib/homestore_backend/heap_chunk_selector.h [hunk 15]
```diff
@@ -35,9 +35,10 @@ class HeapChunkSelector : public homestore::ChunkSelector {
 
     using VChunkHeap = std::priority_queue< VChunk, std::vector< VChunk >, VChunkComparator >;
     using VChunkDefragHeap = std::priority_queue< VChunk, std::vector< VChunk >, VChunkDefragComparator >;
+    using ChunkIdMap = std::unordered_map < homestore::chunk_num_t, homestore::chunk_num_t >;
```
Comment (JacksonYao287) [inline src/lib/homestore_backend/heap_chunk_selector.h:38]: NIT: add a comment here to explain this is used for virtual chunk-> physical chunk map

File: src/lib/homestore_backend/heap_chunk_selector.cpp [hunk 16]
```diff
@@ -186,22 +174,151 @@ void HeapChunkSelector::foreach_chunks(std::function< void(csharedChunk&) >&& cb
...
+                            });
+    auto& pdev_heap = most_avail_dev_it->second;
+    if (num_chunk > pdev_heap->size()) {
+        LOGWARNMOD(homeobject, "Pdev has no enough space to create pg {} with num_chunk {}", pg_id, num_chunk);
+        return std::nullopt;
+    }
+    auto vchunk = VChunk(nullptr);
+    auto it = m_per_pg_heap.emplace(pg_id, std::make_shared< ChunkHeap >()).first;
+    auto v2r_vector = m_v2r_chunk_map.emplace(pg_id, std::make_shared< std::vector < chunk_num_t > >()).first->second;
+    auto r2v_map = m_r2v_chunk_map.emplace(pg_id, std::make_shared< ChunkIdMap >()).first->second;
```
Comment (JacksonYao287) [inline src/lib/homestore_backend/heap_chunk_selector.cpp:230]: pls see the bimap , and check whether it is helpful for the case here?
https://www.boost.org/doc/libs/1_85_0/libs/bimap/doc/html/index.html

File: src/lib/homestore_backend/heap_chunk_selector.h [hunk 17]
```diff
@@ -112,12 +132,18 @@ class HeapChunkSelector : public homestore::ChunkSelector {
...
     uint32_t total_chunks() const;
 
+    uint32_t get_chunk_size() const;
+
 private:
-    std::unordered_map< uint32_t, std::shared_ptr< PerDevHeap > > m_per_dev_heap;
+    std::unordered_map< uint32_t, std::shared_ptr< ChunkHeap > > m_per_dev_heap;
+    std::unordered_map< pg_id_t, std::shared_ptr< ChunkHeap > > m_per_pg_heap;
+    std::unordered_map< pg_id_t, std::shared_ptr< std::vector <chunk_num_t> > > m_v2r_chunk_map;
+    std::unordered_map< pg_id_t, std::shared_ptr< ChunkIdMap > > m_r2v_chunk_map;
```
Comment (JacksonYao287) [inline src/lib/homestore_backend/heap_chunk_selector.h:141]: add comments here to explain the two map.

File: src/lib/homestore_backend/hs_homeobject.hpp [hunk 18]
```diff
@@ -78,15 +78,17 @@ class HSHomeObject : public HomeObjectImpl {
     struct pg_info_superblk {
         pg_id_t id;
         uint32_t num_members;
+        homestore::chunk_num_t num_chunks;
         peer_id_t replica_set_uuid;
+        u_int64_t pg_size;
```
[the remark this answers is not in the document]
Comment (Hooper9973) [reply src/lib/homestore_backend/hs_homeobject.hpp]: done

File: src/lib/homestore_backend/hs_homeobject.hpp [hunk 19]
```diff
@@ -78,15 +78,17 @@ class HSHomeObject : public HomeObjectImpl {
     struct pg_info_superblk {
         pg_id_t id;
         uint32_t num_members;
+        homestore::chunk_num_t num_chunks;
```
[the remark this answers is not in the document]
Comment (Hooper9973) [reply src/lib/homestore_backend/hs_homeobject.hpp]: done

File: src/lib/homestore_backend/hs_pg_manager.cpp (hunk 13 above)
Comment (Hooper9973) [reply src/lib/homestore_backend/hs_pg_manager.cpp]: This ensures that the actual size occupied by the pg is smaller than the specified size, allowing pg moves on machines with sufficient remaining space. I agree with this excellent point.

File: src/lib/homestore_backend/heap_chunk_selector.cpp (hunk 14 above)
Comment (Hooper9973) [reply src/lib/homestore_backend/heap_chunk_selector.cpp:62]: This is a temporary method for shard creation where the pg allocates chunks from its own heap. The final method will be refined by Identical layout.

File: src/lib/homestore_backend/heap_chunk_selector.h (hunk 15 above)
Comment (Hooper9973) [reply src/lib/homestore_backend/heap_chunk_selector.h]: done

File: src/lib/homestore_backend/heap_chunk_selector.cpp (hunk 16 above)
Comment (Hooper9973) [reply src/lib/homestore_backend/heap_chunk_selector.cpp:230]: The underlying implementation of bimap consists of two maps. The current implementation uses a vector to map v_chunk_id to r_chunk_id, ensuring consistency with the order of the persisted superblk and providing O(1) query time complexity. Using bimap might introduce trade-offs to maintain order or enhance querying.

File: src/lib/homestore_backend/heap_chunk_selector.h (hunk 17 above)
Comment (Hooper9973) [reply src/lib/homestore_backend/heap_chunk_selector.h]: done

File: src/lib/homestore_backend/heap_chunk_selector.cpp (hunk 14 above)
Comment (JacksonYao287) [reply src/lib/homestore_backend/heap_chunk_selector.cpp]: pls add a FIXME here

File: src/lib/homestore_backend/replication_state_machine.cpp [hunk 20]
```diff
@@ -121,17 +121,10 @@ ReplicationStateMachine::get_blk_alloc_hints(sisl::blob const& header, uint32_t
...
-        } else if (!shards_found) {
-            // pg is empty without any shards, we leave the decision the HeapChunkSelector to select a pdev
-            // with most available space and then select one chunk based on that pdev
-        } else {
-            return home_object_->chunk_selector()->chunk_to_hints(chunk_id);
-        }
-        break;
+        // Since chunks are selected when a pg is created, the chunkselector selects one of the chunks owned by the pg
+        homestore::blk_alloc_hints hints;
+        hints.pdev_id_hint = msg_header->pg_id; // Temporary bypass using pdev_id_hint to represent pg_id_hint, "identical layout" will change it
```
[the remark this answers is not in the document]
Comment (Hooper9973) [reply src/lib/homestore_backend/replication_state_machine.cpp]: donedone

File: src/lib/homestore_backend/heap_chunk_selector.cpp [hunk 21]
```diff
@@ -186,22 +174,151 @@ void HeapChunkSelector::foreach_chunks(std::function< void(csharedChunk&) >&& cb
...
+
+    const auto chunk_size = get_chunk_size();
+    const uint32_t num_chunk = sisl::round_down(pg_size, chunk_size) / chunk_size;
+
+    //Select a pdev with the most available blocks for it referenced from HeapChunkSelector::select_chunk"
+    auto &&most_avail_dev_it =
+        std::max_element(m_per_dev_heap.begin(), m_per_dev_heap.end(),
+                            [](const std::pair< const uint32_t, std::shared_ptr< ChunkHeap > >& lhs,
+                            const std::pair< const uint32_t, std::shared_ptr< ChunkHeap > >& rhs) {
+                                return lhs.second->available_blk_count.load() < rhs.second->available_blk_count.load();
```
[the remark this answers is not in the document]
Comment (Hooper9973) [reply src/lib/homestore_backend/heap_chunk_selector.cpp]: That makes sense, done.

File: src/lib/homestore_backend/heap_chunk_selector.cpp [hunk 22]
```diff
@@ -186,22 +174,151 @@ void HeapChunkSelector::foreach_chunks(std::function< void(csharedChunk&) >&& cb
...
+
+    const auto& chunk = m_chunks[chunkID];
+    VChunk vchunk(chunk);
+    {
+        std::lock_guard< std::mutex > l(pg_it->second->mtx);
+        auto& pg_heap = pg_it->second->m_heap;
+        pg_heap.emplace(chunk);
+    }
+    auto& avalableBlkCounter = pg_it->second->available_blk_count;
+    avalableBlkCounter.fetch_add(vchunk.available_blks());
```
[the remark this answers is not in the document]
Comment (Hooper9973) [reply src/lib/homestore_backend/heap_chunk_selector.cpp]: done

**Review by xiaoxichen — APPROVED** [review]
lgtm.

I think it will be nice to have a create_pg test that consumes all ( left_chunk < pg_size) chunks.  That might surfaces some issue.

**Review by JacksonYao287 — COMMENTED** [review]
generally looks good. I have a final question.

m_per_dev_heap is now a heap to maintain the chunks of all the pdev , right?
in our implementation, what we want is to mark a chunk has been assigned to any pg or not, so  do we still need a heap here? maybe we can used a simpler data structure?

if I miss something, pls correct me

**Comment by Hooper9973** [comment]
@JacksonYao287 

> m_per_dev_heap is now a heap to maintain all chunks of a pdev , right?
Thanks for bringing this up! Just to clarify a bit more: m_chunks holds all the chunks. m_per_dev_heap specifically manages only the empty chunks, and m_per_pg_heap takes care of the chunks used for specific page groups.

It's important to note that m_per_dev_heap and m_per_pg_heap are mutually exclusive - a chunk will only exist in one or the other, never both. Together, the total number of chunks in m_per_dev_heap and m_per_pg_heap matches exactly with the total in m_chunks.

**Pull request merged**